sandbox every test in its own anonymous module

vtjnash opened the pull request:
this sandboxes each test into its own anonymous module to reduce the likelihood of issues like
https://travis-ci.org/JuliaLang/julia/jobs/47195295

(waiting for CI to give it's approval to merge)

(edit: also need to update replcompletions test)

vtjnash committed:
sandbox every test in its own anonymous module

also fix a bug with eval scope in docs that was being masked by
a named eval-import in the previous structure of that test

use the new f.env.module field in the base serializer also

fix a discrepancy between a comment in methodswith and its implementation

tkelman commented:
also if this didn't break `make testall1`,
![image](https://cloud.githubusercontent.com/assets/5934628/5772714/ec7a5366-9d0b-11e4-9148-3bd58f918692.png)

vtjnash committed:
fix parallel test to work inside a module

need to have a dummy module on all processors so that lambda.code.module
can deserialize successfully

also, Expr(:localize) doesn't work in module-scope due to GlobalRef nodes

vtjnash committed:
also fix the extended parallel tests

vtjnash commented:
replaced by #17435
